src/utils/utils_concurrency.h

```diff
@@ -92,57 +101,148 @@ static __inline unsigned char utils_mssb_index(long long value) {
+    __atomic_load(ptr, out, memory_order_acquire);
+    //utils_annotate_acquire(ptr);
+}
 
-#define utils_atomic_store_release(object, desired)                            \
-    do {                                                                       \
-        __atomic_store_n(object, desired, memory_order_release);               \
-        utils_annotate_release((void *)object);                                \
-    } while (0)
+static inline void utils_atomic_load_acquire_ptr(void **ptr, void **out) {
+    ASSERT_IS_ALIGNED((uintptr_t)ptr, 8);
+    ASSERT_IS_ALIGNED((uintptr_t)out, 8);
+    __atomic_load((uintptr_t *)ptr, (uintptr_t *)out, memory_order_acquire);
+    //utils_annotate_acquire((void *)ptr);
+}
 
-#define utils_atomic_increment(object)                                         \
-    __atomic_add_fetch(object, 1, memory_order_acq_rel)
+static inline void utils_atomic_store_release_ptr(void **ptr, void *val) {
+    ASSERT_IS_ALIGNED((uintptr_t)ptr, 8);
+    //utils_annotate_release(ptr);
+    __atomic_store_n((uintptr_t *)ptr, (uintptr_t)val, memory_order_release);
+}
 
-#define utils_atomic_decrement(object)                                         \
-    __atomic_sub_fetch(object, 1, memory_order_acq_rel)
+static inline uint64_t utils_atomic_increment_u64(uint64_t *val) {
+    ASSERT_IS_ALIGNED((uintptr_t)val, 8);
+    // return incremented value
+    return __atomic_add_fetch(val, 1, memory_order_acq_rel);
+}
 
-#define utils_fetch_and_add64(object, value)                                   \
-    __atomic_fetch_add(object, value, memory_order_acq_rel)
+static inline uint64_t utils_atomic_decrement_u64(uint64_t *val) {
+    ASSERT_IS_ALIGNED((uintptr_t)val, 8);
+    // return decremented value
+    return __atomic_sub_fetch(val, 1, memory_order_acq_rel);
+}
 
-#define utils_compare_exchange(object, expected, desired)                      \
-    __atomic_compare_exchange(object, expected, desired, 0 /* strong */,       \
-                              memory_order_acq_rel, memory_order_relaxed)
+static inline uint64_t utils_fetch_and_add_u64(uint64_t *ptr, uint64_t val) {
+    ASSERT_IS_ALIGNED((uintptr_t)ptr, 8);
+    // return the value that had previously been in *ptr
+    return __atomic_fetch_add(ptr, val, memory_order_acq_rel);
+}
+
+static inline uint64_t utils_fetch_and_sub_u64(uint64_t *ptr, uint64_t val) {
+    // return the value that had previously been in *ptr
+    ASSERT_IS_ALIGNED((uintptr_t)ptr, 8);
+    return __atomic_fetch_sub(ptr, val, memory_order_acq_rel);
+}
+
+static inline bool utils_compare_exchange_u64(uint64_t *ptr, uint64_t *expected,
+                                              uint64_t *desired) {
+    ASSERT_IS_ALIGNED((uintptr_t)ptr, 8);
+    ASSERT_IS_ALIGNED((uintptr_t)expected, 8);
+    ASSERT_IS_ALIGNED((uintptr_t)desired, 8);
+
+    // if (*ptr == *expected)
+    //   *ptr = *desired
+    //   return true
+    // else
+    //  *expected = *ptr
+    // return false
```

Comment:
I do not think this comment is needed, because everyone knows or can google how the `__atomic_compare_exchange` works.

Reply:
removed